File: DevelopArchitecture/DevelopArchitecture/architecture/loginUserInfo/JLoginUserModel.m
//
//  JLoginUserModel.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 15/1/23.
//  Copyright (c) 2015年 Jeffrey. All rights reserved.
//

#import "JLoginUserModel.h"
#import "JConstantHelp.h"
#import "JAppUtility.h"
#import "JOperationDB.h"
#import "NSDictionary+JSON.h"

static NSMutableDictionary * userDic =nil;
@interface JLoginUserModel()
@end
@implementation JLoginUserModel

+(void)funj_setUserDicData:(BOOL)isReload{
    if(!userDic){
        userDic =[[NSMutableDictionary alloc]init];
    }
    if(isReload){
        [userDic removeAllObjects];
    }
}
//目前只适合 登录与环信登录 分开保存，
//{@"islogining":@"0"}
+(void)funj_insertLoginUserMessage:(NSDictionary*)data{
    if(!data || !data[@"userId"]) return;
    
    NSMutableDictionary *idKeyDic =[[NSMutableDictionary alloc]init];
    for(NSString *name in [self getItems]){
        if([name isEqualToString:@"userId"])continue;
        [idKeyDic setObject:@"text" forKey:name];
    }
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [JOperationDB funj_createTable:kUserLoginDB mainKey:@{@"userId":@"text"}
                         otherKey:idKeyDic];
    });
    NSString *serverTime =[JAppUtility funj_getDateTime:@"yyyyMMddHHmmss"];
    NSMutableDictionary *dic=[[NSMutableDictionary alloc]initWithDictionary:data];
    [dic setObject:serverTime forKey:@"serverTime"];
    
    [JOperationDB funj_insertData:kUserLoginDB data:dic mainKey:@{@"userId":[dic objectForKey:@"userId"]}];
    
    [self funj_setUserDicData:YES];
    
}
+(NSDictionary*)funj_getLastLoginUserMessage{
    [self funj_setUserDicData:NO];
    
    if([userDic count] >0){
        return userDic;
    }
    NSDictionary *dic= [JOperationDB funj_searchTopFirstFromTable:kUserLoginDB :[self getItems] searchKey:@"serverTime"];
    [userDic addEntriesFromDictionary:dic];
    return (dic && [dic count]>0) ? userDic : [NSMutableDictionary dictionary];
}
+(NSMutableDictionary*)funj_getLastLoginTokenMessage{
    [self funj_setUserDicData:NO];
    if( [userDic count]>0){
        NSMutableDictionary *dic =[[NSMutableDictionary alloc]init];
        if([[userDic objectForKey:@"isLogining"]isEqualToString:@"1"]){
            [dic setObject:userDic[@"token"] forKey:@"token"];
        }else{
            [dic setObject:@"" forKey:@"token"];
        }
        return dic;
    }else{
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            [self funj_getLastLoginUserMessage];
        });
    }
    
    NSMutableDictionary*dic = (NSMutableDictionary*)[JOperationDB funj_searchTopFirstFromTable:kUserLoginDB :@[@"token",@"isLogining"] searchKey:@"serverTime"];
    if(dic && [dic count]>0){
        if(![[dic objectForKey:@"isLogining"]isEqualToString:@"1"]){
            [dic setObject:@"" forKey:@"token"];
        }
        [dic removeObjectForKey:@"isLogining"];
    }
    
    return dic;
}
+(BOOL)funj_getIsLogining{
    NSDictionary *dic = [self funj_getLastLoginUserMessage];
    return [dic integerWithKey:@"isLogining"] == 1;
}
+(void)funj_deleteUserMessage:(NSDictionary*)data{
    [JOperationDB funj_deleteDataFromTable:kUserLoginDB deleteKey:data];
    [self funj_setUserDicData:YES];
}
+(NSArray*)getItems{
    return @[@"userId",@"photoId",@"nickName",@"mySign",@"password",
             @"serverTime",@"token",@"userType",@"userName",@"cloudId",
             @"isLogining",@"phone",@"account",@"isaccountlogin",@"originalPassCode",
             @"schoolName",@"sex",@"schoolId",@"photoPath",@"cloud_usr_account",@"relPhone"];
}

@end

File: DeveloperArchitecture/DeveloperArchitecture/architecture/loginitem/JLoginUserModel.swift
//
//  JLoginUserModel.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/9/29.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit
import CoreData

public class JLoginUserModel : JBaseDataModel {
    static var m_userModel : UserLoginDB? = nil

    class func funj_insertLoginUserMessage(_ data : Dictionary<String,Any>) {
        if data["userId"] == nil || data.count <= 0 {
            return
        }
        
        let properyList = UserLoginDB.funj_allPropertyList()
        var dict : Dictionary<String,String> = Dictionary()
        for (k,v) in data {
            if properyList.contains(k) {
                dict[k] = "\(v)"
            }
        }
        let serverTime = JAppUtility.funj_getDateTime(nil)
        dict["serverTime"] = serverTime
        
        JCoreDataManager.shared.funj_insertMessageContext(UserLoginDB.fetchRequest(), dict, mainDic: ["userId":dict["userId"]!])
        self.funj_setUserDicData(true)
    }
    class func funj_getLastLoginUserMessage() -> UserLoginDB? {
        var loginModel : UserLoginDB? = nil
        if m_userModel != nil {
            loginModel = m_userModel
        }else {
            let userList = JCoreDataManager.shared.funj_getMessageContext(UserLoginDB.fetchRequest(), dict: nil, asc: nil, desc: "serverTime", limit: 1)
            
            if userList.count > 0 {
                loginModel = userList.first as? UserLoginDB
            }
        }
        
        return loginModel
    }
    class func funj_getLastLoginTokenMessage() -> String {
        var token : String = ""
        let loginModel = JLoginUserModel.funj_getLastLoginUserMessage()
        token = loginModel?.token ?? ""
        return token
    }
    class func funj_getIsLogining() -> Bool {
        var isLogining : Bool = false
        let loginModel = JLoginUserModel.funj_getLastLoginUserMessage()
        isLogining = loginModel?.isLogining == "1" ? true : false
        return isLogining
    }
    class func funj_deleteUserMessage(_ userId : String) {
        JCoreDataManager.shared.funj_deleteMessageContext(UserLoginDB.fetchRequest(), dict: ["userId":userId])
        self.funj_setUserDicData(true)
    }
    class func funj_setUserDicData(_ reload : Bool){
        if reload {
            m_userModel = nil
        }
    }
}

@objc(UserLoginDB)
public class UserLoginDB : NSManagedObject {
    public override class func fetchRequest() -> NSFetchRequest<NSFetchRequestResult> {
        return NSFetchRequest<NSFetchRequestResult>(entityName: "UserLoginDB")
    }
    
    @NSManaged public var account: String?
    @NSManaged public var isLogining: String?
    @NSManaged public var mySign: String?
    @NSManaged public var nickName: String?
    @NSManaged public var sex: String?
    @NSManaged public var originalPassCode: String?
    @NSManaged public var password: String?
    @NSManaged public var photoPath: String?
    @NSManaged public var serverTime: String?
    @NSManaged public var token: String?
    @NSManaged public var userId: String?
    @NSManaged public var userName: String?
}
